File: app/src/main/java/com/bongorian/signa1/AboutDialog.java
package com.bongorian.signa1;

import android.app.AlertDialog;
import android.content.Intent;
import android.net.Uri;
import android.text.util.Linkify;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.LinearLayout;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;

/** Privacy information remains available offline and without camera permission. */
final class AboutDialog {
    static void show(MainActivity activity){
        String policy;
        try(InputStream in=activity.getResources().openRawResource(R.raw.privacy_policy);ByteArrayOutputStream out=new ByteArrayOutputStream()){
            byte[] buffer=new byte[4096];int n;while((n=in.read(buffer))!=-1)out.write(buffer,0,n);
            policy=out.toString(StandardCharsets.UTF_8.name());
        }catch(Exception error){policy=activity.getString(R.string.ui_could_not_load_the_privacy_policy_contact_dennosamurai);}
        TextView text=activity.text(policy,14,MainActivity.WHITE);text.setTextIsSelectable(true);
        text.setPadding(activity.dp(20),activity.dp(12),activity.dp(20),activity.dp(16));text.setLineSpacing(activity.dp(4),1);
        text.setAutoLinkMask(Linkify.EMAIL_ADDRESSES);text.setLinkTextColor(MainActivity.LIME);
        LinearLayout content=new LinearLayout(activity);content.setOrientation(LinearLayout.VERTICAL);
        TextView licenses=activity.button(activity.getString(R.string.ui_open_source_licenses));licenses.setOnClickListener(v->showLicenses(activity));
        content.addView(licenses,new LinearLayout.LayoutParams(-1,activity.dp(48)));content.addView(text);
        ScrollView scroll=new ScrollView(activity);scroll.addView(content);
        new AlertDialog.Builder(activity).setTitle(BuildConfig.APP_NAME+" · "+BuildConfig.VERSION_NAME)
            .setView(scroll).setPositiveButton(activity.getString(R.string.ui_close),null)
            .setNeutralButton(activity.getString(R.string.ui_contact),(dialog,which)->{
                Intent intent=new Intent(Intent.ACTION_SENDTO,Uri.parse("mailto:dennosamurai@gmail.com"));
                intent.putExtra(Intent.EXTRA_SUBJECT,"5igna1 support");
                try{activity.startActivity(intent);}catch(Exception unavailable){android.widget.Toast.makeText(activity,activity.getString(R.string.ui_contact_dennosamurai_gmail_com),android.widget.Toast.LENGTH_LONG).show();}
            }).show();
    }
    static void showLicenses(MainActivity activity){
        StringBuilder legal=new StringBuilder();
        for(String name:new String[]{"NOTICE","THIRD_PARTY_LICENSES.md","LICENSE","Kotlin-COPYRIGHT.txt","Kotlin-Boost-1.0.txt"}){
            try(InputStream in=activity.getAssets().open("licenses/"+name);ByteArrayOutputStream out=new ByteArrayOutputStream()){
                byte[] buffer=new byte[4096];int n;while((n=in.read(buffer))!=-1)out.write(buffer,0,n);
                legal.append(name).append("\n\n").append(out.toString(StandardCharsets.UTF_8.name())).append("\n\n");
            }catch(Exception error){legal.append(name).append(": ").append(activity.getString(R.string.ui_unavailable)).append('\n');}
        }
        TextView text=activity.text(legal.toString(),12,MainActivity.WHITE);text.setTextIsSelectable(true);text.setPadding(activity.dp(20),activity.dp(12),activity.dp(20),activity.dp(16));
        ScrollView scroll=new ScrollView(activity);scroll.addView(text);
        new AlertDialog.Builder(activity).setTitle(activity.getString(R.string.ui_open_source_licenses)).setView(scroll).setPositiveButton(activity.getString(R.string.ui_close),null).show();
    }
}

File: app/src/main/kotlin/com/bongorian/signa1/AboutDialog.kt
package com.bongorian.signa1

import android.content.Intent
import android.net.Uri
import android.text.util.Linkify
import android.view.View
import android.widget.LinearLayout
import android.widget.Toast
import java.io.ByteArrayOutputStream
import java.nio.charset.StandardCharsets

/** Privacy information remains available offline and without camera permission. */
internal object AboutDialog {
    fun show(activity: MainActivity) {
        var policy: String?
        try {
            activity.resources.openRawResource(R.raw.privacy_policy).use { `in` ->
                ByteArrayOutputStream().use { out ->
                    val buffer = ByteArray(4096)
                    var n: Int
                    while ((`in`.read(buffer).also { n = it }) != -1) out.write(buffer, 0, n)
                    policy = out.toString(StandardCharsets.UTF_8.name())
                }
            }
        } catch (error: Exception) {
            policy =
                activity.getString(
                    R.string.ui_could_not_load_the_privacy_policy_contact_dennosamurai
                )
        }
        val text = activity.text(policy, 14, MainActivity.WHITE)
        text.setTextIsSelectable(true)
        text.setPadding(activity.dp(20f), activity.dp(12f), activity.dp(20f), activity.dp(16f))
        text.setLineSpacing(activity.dp(4f).toFloat(), 1f)
        text.setAutoLinkMask(Linkify.EMAIL_ADDRESSES)
        text.setLinkTextColor(MainActivity.LIME)
        val content = LinearLayout(activity)
        content.setOrientation(LinearLayout.VERTICAL)
        val licenses = activity.button(activity.getString(R.string.ui_open_source_licenses))
        licenses.setOnClickListener(OnClickListener@{ v: View? -> showLicenses(activity) })
        content.addView(licenses, LinearLayout.LayoutParams(-1, activity.dp(48f)))
        content.addView(text)
        SignalSheet.content(
            activity,
            BuildConfig.APP_NAME + " · " + BuildConfig.VERSION_NAME,
            content,
            R.string.ui_contact,
            Runnable {
                val intent =
                    Intent(Intent.ACTION_SENDTO, Uri.parse("mailto:dennosamurai@gmail.com"))
                intent.putExtra(Intent.EXTRA_SUBJECT, "5igna1 support")
                try {
                    activity.startActivity(intent)
                } catch (unavailable: Exception) {
                    Toast.makeText(
                            activity,
                            activity.getString(R.string.ui_contact_dennosamurai_gmail_com),
                            Toast.LENGTH_LONG,
                        )
                        .show()
                }
            },
            .87f,
        )
    }

    fun showLicenses(activity: MainActivity) {
        val legal = StringBuilder()
        for (name in
            arrayOf<String>(
                "NOTICE",
                "THIRD_PARTY_LICENSES.md",
                "LICENSE",
                "Kotlin-COPYRIGHT.txt",
                "Kotlin-Boost-1.0.txt",
            )) {
            try {
                activity.assets.open("licenses/" + name).use { `in` ->
                    ByteArrayOutputStream().use { out ->
                        val buffer = ByteArray(4096)
                        var n: Int
                        while ((`in`.read(buffer).also { n = it }) != -1) out.write(buffer, 0, n)
                        legal
                            .append(name)
                            .append("\n\n")
                            .append(out.toString(StandardCharsets.UTF_8.name()))
                            .append("\n\n")
                    }
                }
            } catch (error: Exception) {
                legal
                    .append(name)
                    .append(": ")
                    .append(activity.getString(R.string.ui_unavailable))
                    .append('\n')
            }
        }
        val text = activity.text(legal.toString(), 12, MainActivity.WHITE)
        text.setTextIsSelectable(true)
        text.setPadding(activity.dp(20f), activity.dp(12f), activity.dp(20f), activity.dp(16f))
        SignalSheet.content(
            activity,
            activity.getString(R.string.ui_open_source_licenses),
            text,
            0,
            null,
            .87f,
        )
    }
}
